napari/plugins/tests/test_hookspecs.py

```diff
@@ -0,0 +1,77 @@
+import inspect
+import warnings
+
+import pytest
+from numpydoc.docscrape import FunctionDoc
+
+from napari.plugins import hookspecs
+
+HOOKSPECS = []
+for name, func in vars(hookspecs).items():
+    if hasattr(func, 'napari_spec'):
+        HOOKSPECS.append((name, func))
```

Comment:
Would you mind explaining napari.plugins.hookspecs in a comment here? I'm still confused about how hookspecs work their magic. Or even what they are 😂. The more places they are described, the better, I think.

Reply:
hah.  without duplicating the [pluggy docs](https://pluggy.readthedocs.io/en/latest/#how-does-it-work), the basic idea here is this:
1. we first create a hookspec decorator:
   `hookspec = pluggy. HookspecMarker("project_name")`
2. when it decorates a function, that function object gets a new attribute called "`project_name_spec`"
3. that attribute is what makes them discoverable when you run `plugin_manager.add_hookspecs(module)`
4. on the plugin side, they create a new decorator (or import it from us): 
   `hookimpl = pluggy.HookimplMarker("project_name")`
5. when it decorates a function, that function object gets a new attribute called "`project_name_impl`"
6. that attribute is what makes them discoverable when you run `plugin_manager.register(plugin_module)`
7. during registration, hookimpls get matched up with hookspecs of the same name.
8. finally, in our code, when we call `plugin_manager.hook.name_of_hookspec(...)`, pluggy takes care of calling the hookimpl for all registered functions and returning the results

a hookspec itself is nothing more than a function signature.  just contract that we intend to implement somewhere in our code.  the `hookimpl` satisfies that contract.

Reply:
Ok, so yesterday I was just too tired for the above to register. Today it's making a bit more sense, but I still wouldn't say it's crisp for me.

The first thing is the word "hook", I think we can do better. Can "hook" be accurately translated as "pluggable"?

I also kinda hate the abbreviation of everything to four letters. If I replace "hookimpl" with "pluggable_implementation" (parsed as "implementation of a pluggable"), things are starting to make more sense. Maybe instead of `hookimpl` we can use `napari_plugin_implementation` as our decorator?

Similarly, I'd like to replace our use of `hookspec` with `plugin_specification`. Of course, we can provide a guide in our documentation for how our terms translate to pluggy's, but I really do dislike the pluggy vocabulary.

Finally, I guess "project_name" is the namespace in which plugin implementations get matched to specification? Should we make multiple napari namespaces (io, interactive, functional) in order to prevent future plugin clashes? I can imagine cases of functional plugins clashing with IO plugins, for example...

Anyway, things to discuss in meeting in a few mins! =)

Reply:
As per discussion let's go with `hook_implementation` and `hook_specification` in our usage and make it clear in documentation that these correspond to `hookimpl` and `hookspec` within pluggy

Reply:
Please don't use different names for these, it'll make working out what's what using pluggy's documentation a lot harder!

Reply:
There was quite extensive discussion around this just now in our team meeting. I'll try and capture some of that here. I think the major con as you note it that there might be some confusion with respect to pluggy documentation, but it was also noted that these are really just arbitrary variable names that only need to be consistent within a file and that are independent of whatever names plugin developers choose to use within their plugins (which we have no control of). For example, we could make them `xyz` and `uvw` and the plugin ecosystem would work just the same, the only thing that really changes is the readability and understandability of our own code. When it comes to a comparison with a pluggy documentation example, what we are planning to write is

```python
hook_specification = pluggy.HookspecMarker("myproject")
hook_implentation = pluggy.HookimplMarker("myproject")
```
instead of 
```python
hookspec = pluggy.HookspecMarker("myproject")
hookimpl = pluggy.HookimplMarker("myproject")
```
because we considered that more readable and more consistent with the rest of the napari code base, but both of these would work just fine. In other places in napari (say around the vispy code) we also choose to use naming conventions that are consistent with the rest of napari rather than the underlying library and I think that makes napari overall much more readable. There also been some thought on how to do this best with out qtpy code as well as they have quite different naming conventions too, see #6. 

We also don't expect most plugin developers to have to consult the pluggy documentation for making plugins - we plan to explain the majority of what they need in our docs. For the smaller number that do consult the pluggy docs hopefully what we are doing will be clear.

Is this reasoning compelling to you @GenevieveBuckley ? There are certainly people on both sides of this discussion, but I hope we've reached a decision

Reply:
I think we all want the same thing, which is for people to be less (not more) confused. Sorry about posting at almost the same time as your meeting!

Reply:
You're fine @GenevieveBuckley - also if you'd like to join our team meeting they are at 4pm PST Thursdays - @jni can give you the details in the Zullip and you're more than welcome.

I'm going to be signing off for holiday soon, so I'll trust it to @tlambert03 and @jni make a final call on this one :-)

Reply:
Happy holidays @sofroniewn!